### PR TITLE
Added blocklist by title / questID

### PR DESCRIPTION
**Pull request by zgavin**
Hey there @p3lim.  Thanks for continuing to work on this addon, it's saved me a ton of time over the years.   I implemented this title / questID based blocklist for myself midway through BFA and figured I would share it now that you're pushing out frequent updates again.  Also has a small bug fix for using the enter key in input boxes for quest items/npcs.  Hope this meets your approval.

### CODE REVIEW

**Comment by p3lim** [comment]
Thanks for the contribution, I'll give it a rudamentary look now before testing it, but why is this needed vs item/npc blocklists? Just for granularity? Or are there cases where the existing blocklists don't work?

File: Addon.lua [hunk 1]
```diff
@@ -32,6 +32,22 @@ local rogueNPCs = {
...
 }
 
+local function IsQuestIgnored( questID )
+	if ignoredQuests[questID] then
+		return true
+	end
+
+	local title = C_QuestLog.GetTitleForQuestID(questID)
+
+	for _, titleOrQuestID in ipairs(ns.db.profile.blocklist.title) do
```
Comment (p3lim) [inline Addon.lua:42]: Use `next`, not `ipairs`.

File: Addon.lua [hunk 2]
```diff
@@ -32,6 +32,22 @@ local rogueNPCs = {
 	[93188] = true, -- Mongar
 }
 
+local function IsQuestIgnored( questID )
+	if ignoredQuests[questID] then
+		return true
+	end
+
+	local title = C_QuestLog.GetTitleForQuestID(questID)
+
```
Comment (p3lim) [inline Addon.lua:41]: Remove space.

File: config/OptionsBlocklist.lua [hunk 3]
```diff
@@ -222,9 +222,81 @@ local function CreateNPCBlocklistOptions()
 	end)
 end
 
+local function CreateTitleBlocklistOptions()
+	local panel = CreateOptionsPanel('TitleBlocklist',
+		L['Title Blocklist'],
+		L['Quests with titles that partially match or IDs that exactly match entries from this list will not be automated.'],
+		L['Block Title'])
```
Comment (p3lim) [inline config/OptionsBlocklist.lua:229]: These three should be added as commented out lines in the various localization files.

File: config/OptionsBlocklist.lua [hunk 4]
```diff
@@ -222,9 +222,81 @@ local function CreateNPCBlocklistOptions()
...
+
+	local function OnRemove(self)
+		for i, title in ipairs(ns.db.profile.blocklist.title) do
```
Comment (p3lim) [inline config/OptionsBlocklist.lua:232]: Use `next`, not `ipairs`.

File: config/OptionsBlocklist.lua [hunk 5]
```diff
@@ -222,9 +222,81 @@ local function CreateNPCBlocklistOptions()
...
+			if title == self.title then
+				tremove(ns.db.profile.blocklist.title, i)
```
Comment (p3lim) [inline config/OptionsBlocklist.lua:234]: Use `ns.db.profile.blocklist.title[i] = nil`.

File: config/OptionsBlocklist.lua [hunk 6]
```diff
@@ -222,9 +222,81 @@ local function CreateNPCBlocklistOptions()
...
+			end
+		end
+	end
+	
+	local titles = ns.db.profile.blocklist.title
```
Comment (p3lim) [inline config/OptionsBlocklist.lua:239]: Don't upvalue this.

File: config/SavedVariables.lua [hunk 7]
```diff
@@ -156,6 +157,10 @@ ns.EventHandler:Register('ADDON_LOADED', function(...)
 			QuickQuestBlacklistDB = nil
 		end
 
+		if not ns.db.profile.blocklist.title then
+			ns.db.profile.blocklist.title = {}
+		end
```
Comment (p3lim) [inline config/SavedVariables.lua:162]: AceDB should be able to merge this by itself, let me know if it doesn't.

File: config/OptionsPopups.lua [hunk 8]
```diff
@@ -6,8 +6,8 @@ StaticPopupDialogs[addonName .. 'ItemBlocklistPopup'] = {
 	button1 = L['Accept'],
 	button2 = L['Cancel'],
 	hasEditBox = true,
-	EditBoxOnEnterPressed = function(self)
-		self.data.callback(self.data.pool, tonumber(strtrim(self.editBox:GetText())))
+	EditBoxOnEnterPressed = function(self, data)
+		data.callback(data.pool, tonumber(strtrim(self:GetText())))
```
Comment (p3lim) [inline config/OptionsPopups.lua:10]: I'd prefer this fix in a separate PR.

**Comment by zgavin** [comment]
> why is this needed vs item/npc blocklists? Just for granularity?

Yep, pretty much.  There were two specific examples that lead me to create this:

* Emissaries and paragon chests get turned in to the same NPCs. At some points in BFA it was better to bank paragon chests, but I still wanted to automate emissaries.
* The NPC for conquest turn in and trophy turn in were the same.  Like the paragons, sometimes it was best to bank your trophy turn ins (eg, if you expected your rating to increase).

I'll make the changes you requested and push up later today.

I'll also push up the edit box fix as a separate.

File: config/SavedVariables.lua (hunk 7 above)
Comment (zgavin) [reply config/SavedVariables.lua:162]: cool I wasn't sure what ace's behavior was 👍

File: config/OptionsBlocklist.lua (hunk 5 above)
Comment (zgavin) [reply config/OptionsBlocklist.lua:234]: This would leave a gap in the array that would need to be guarded against elsewhere.

**Comment by zgavin** [comment]
@p3lim I've updated the branch as you requested.  The only change i didn't make was to change `tremove` to `... = nil`, as it breaks the `tContains` / `tinsert` later (because they expect sequential arrays).   I'm happy to update this further if you feel strongly, but wasn't sure what you'd prefer.

**Comment by p3lim** [comment]
> The only change i didn't make was to change tremove to ... = nil, as it breaks the tContains / tinsert later (because they expect sequential arrays).
That one was fair.

I'm gonna give this an actual test this week and come back to you.

**Comment by p3lim** [comment]
Needs a little tweaking, but looks good.

**Comment by p3lim** [comment]
Feel free to suggest more quests to be added to the default list:
https://github.com/p3lim-wow/QuickQuest/blob/a217dd8ae8c867b0da9ddefc2a6c77c1dd97f00a/config/SavedVariables.lua#L95

**Pull request merged**